File: week_3_dp/ts-dp/singleton.ts
// 餓漢式
class Singleton1 {
  // 1. 構造器私有化，外部不能new
  private constructor() {}
  // 2. 本類內部創建物件實例化
  private static instance = new Singleton1();

  public static getInstance() {
    return this.instance;
  }
}
//console.log(Singleton1.getInstance(), "1111");

// 懶漢式
class Singleton2 {
  private constructor() {}
  private static instance: Singleton2 | null = null;
  public static getInstance() {
    if (this.instance === null) {
      this.instance = new Singleton2();
    }
    return this.instance;
  }
}
//console.log(Singleton2.getInstance(), "2222");

enum Singleton {
  INSTANCE,
}

const instance = Singleton.INSTANCE;
const instance1 = Singleton.INSTANCE;
// 枚舉是單例
console.log(instance === instance1);

File: week_3_dp/Java-DesignPattern/src/main/java/com/pattern/singleton/demo2/Singleton.java
package com.pattern.singleton.demo2;

// 餓漢式: 靜態代碼塊
public class Singleton {
    // 私有構造方法
    private Singleton() {
    }

    // 聲明 Singleton 類型的變量
    private static Singleton instance;  // null

    // 在靜態代碼塊中進行賦值
    static {
        instance = new Singleton();
    }

    // 對外提供獲取該類物件的方法
    public static Singleton getInstance() {
        return instance;
    }
}

class Client {
    public static void main(String[] args) {
        Singleton instance = Singleton.getInstance();
        Singleton instance1 = Singleton.getInstance();
        System.out.println(instance == instance1);
    }
}